File: C++/Algorithms/SortingAlgorithms/MergeSort.cpp
#include<iostream>
#include<vector>
using namespace std;
void merge(vector<int>& arr, int l,int m,int r){
    int n1=m-l+1,n2=r-m;
    int left[n1],right[n2];
    for(int i=0;i<n1;i++){
        left[i]=arr[l+i];
    }
    for(int j=0;j<n2;j++){
        right[j]=arr[m+j+1];
    }

    int i=0,j=0,k=l;
    while(i<n1 && j<n2){
        if(left[i]<=right[j]){
            arr[k]=left[i];
            i++;
            k++;
        }
        else{
            arr[k]=right[j];
            j++;
            k++;
        }
        

    }
    while(i<n1){
        arr[k]=left[i];
        i++;
        k++;
    }
    while(j<n2){
        arr[k]=right[j];
        k++;
        j++;
    }
}
void merge_sort(vector<int>& arr, int l,int r) {
if(l<r){
    int m=l+(r-l)/2;
    merge_sort(arr,l, m);
    merge_sort(arr,m+1, r);
    merge(arr,l,m,r);
}
}

int main() {
    int size;
    cin>>size;
    vector<int>vec;
    for (int i=0;i<size;i++) {
        int p;
        cin>>p;
        vec.push_back(p);

    }
    merge_sort(vec,0, size-1);
    for (int i=0;i<size;i++) {
        cout<<vec[i]<<" ";
    }
    cout<<endl;
}

File: Python/Algorithms/SortingAlgorithms/MergeSort.py
def merge(self, left, right):
    if len(left) == 0:
        return right

    if len(right) == 0:
        return left

    result = []
    index_left = index_right = 0

    while len(result) < len(left) + len(right):
        if left[index_left] <= right[index_right]:
            result.append(left[index_left])
            index_left += 1
        else:
            result.append(right[index_right])
            index_right += 1

        if index_right == len(right):
            result += left[index_left:]
            break

        if index_left == len(left):
            result += right[index_right:]
            break
    return result


def mergeSort(self, arr):
    """ mergeSort Algorithm Implementation in Python 3

        arr : Unorded list
        output : Return list in ascending order.
        time complexity : O(n log2n)
        Note : O(n log2n) is the best possible worst-case runtime that can be
        achieved by a sorting algorithm.

        Example :
        >>> mergeSort([4,2,6,5,9,8])
        [2, 4, 5, 6, 8, 9]"""

    if len(arr) < 2:
        return arr

    midpoint = len(arr) // 2

    return merge(left=mergeSort(arr[:midpoint]),
                 right=mergeSort(arr[midpoint:]))
